### ISSUE TITLE
what does contextInitialized of ServletContextListener really mean?

### ISSUE THREAD

**Issue by xiaoshuang-lu** [eclipse/jetty.project#2049]
Hi Jetty guys, I am a fresh man of Java servlet. And at present,  I wanna introduce service registry for web containers. As far as I know, address and port of web container shall be registered to a service registry after the web container is ready to accept sockets from clients. Does invocation of contextInitialized mean that the server is ready to handle connection requests? Thanks.

**Comment by gregw** [comment]
@xiaoshuang-lu This is not really an issue and probably should have been asked on `jetty-user` or `jetty-discuss` mailing lists.    But you've asked now, so I will answer here and close.

There is no strict relationship between listening server sockets and contexts. At best the relationship can be described and M:N, where you can have 0..M server sockets accepting requests for 1..N contexts.  But this is a dynamic relationship and server sockets may be opened after contexts are initialised. Some configurations may not even have any server sockets - for example requests might arrive via a unix-domain socket connector or a gRPC connector.

More over, the order of starting connectors and contexts is not defined.  Some installations will not start contexts until they are sure the server socket(s) are open.  Others will not open the server socket(s) until they are sure the contexts have started OK.

I believe a good practise is to use the order: open the server socket(s) but don't start accepting; start the contexts; start accepting on the connectors.   This can be done in Jetty, but is not currently supported by our default configuration.... actually I think I will leave this issue open so this can be fixed.

Finally, another complication is that the `contextInitilized` listener is badly named and it in fact called at the start of initialization rather than at the end. It's javadoc says:
```
Receives notification that the web application initialization process is starting. 
All ServletContextListeners are notified of context initialization before any filters or servlets in the web 
 application are initialized.
```
So in summary,  `contextInitialized`  is not the method you are looking for.  I'm not sure there is a standard way to achieve what you want.

**Comment by gregw** [comment]
Added option to open before starting contexts in #2053

**Comment by xiaoshuang-lu** [comment]
Hi @gregw, would you mind to mark this ticket with questions label or something like that?

**Comment by xiaoshuang-lu** [comment]
@gregw, thank you for your detailed interpretation. I appreciate it.

I will work around with an initialization parameter of ServletContext to manipulate the delay of service registry, since the the order of starting connectors and contexts is not defined.

Again, thanks.
